File: 1373-maximum-sum-bst-in-binary-tree/1373-maximum-sum-bst-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = 0;
    
    vector<int> solve(TreeNode *root) {
        if (!root) return {INT_MAX, INT_MIN, 0};
        vector<int> left = solve(root->left);
        vector<int> right = solve(root->right);
        if (left[1] >= root->val || right[0] <= root->val) {
            return {INT_MIN, INT_MAX, 0};
        }
        int sum = left[2] + right[2] + root->val;
        res = max(res, sum);
        return {min(root->val, left[0]), max(root->val, right[1]), sum};
    }

    int maxSumBST(TreeNode* root) {
        solve(root);
        if (res < 0) return 0;
        return res;
    }
};

File: 1373-maximum-sum-bst-in-binary-tree/1373-maximum-sum-bst-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int res = 0;

    public int[] solve(TreeNode root) {
        if (root == null) {
            return new int[]{Integer.MAX_VALUE, Integer.MIN_VALUE, 0};
        }
        
        int[] left = solve(root.left);
        int[] right = solve(root.right);
        
        if (left[1] >= root.val || right[0] <= root.val) {
            return new int[]{Integer.MIN_VALUE, Integer.MAX_VALUE, 0};
        }
        
        int sum = left[2] + right[2] + root.val;
        res = Math.max(res, sum);
        
        return new int[]{Math.min(root.val, left[0]), Math.max(root.val, right[1]), sum};
    }

    public int maxSumBST(TreeNode root) {
        solve(root);
        if (res < 0) return 0;
        return res;
    }
}